crates/noirc_abi/src/lib.rs

```diff
@@ -97,6 +116,60 @@ pub enum Sign {
 }
 
 impl AbiType {
+    // TODO: Add `Context` argument for resolving fully qualified struct paths
+    pub fn from_type(typ: &Type) -> Self {
+        // Note; use strict_eq instead of partial_eq when comparing field types
+        // in this method, you most likely want to distinguish between public and private
+        match typ {
+            Type::FieldElement => Self::Field,
+            Type::Array(size, typ) => {
+                let length = size
+                    .evaluate_to_u64()
+                    .expect("Cannot have variable sized arrays as a parameter to main");
+                let typ = typ.as_ref();
+                Self::Array { length, typ: Box::new(Self::from_type(typ)) }
+            }
+            Type::Integer(sign, bit_width) => {
+                let sign = match sign {
+                    Signedness::Unsigned => Sign::Unsigned,
+                    Signedness::Signed => Sign::Signed,
+                };
+
+                Self::Integer { sign, width: *bit_width }
+            }
+            Type::TypeVariable(binding, TypeVariableKind::IntegerOrField) => {
+                match &*binding.borrow() {
+                    TypeBinding::Bound(typ) => Self::from_type(typ),
+                    TypeBinding::Unbound(_) => Self::from_type(&Type::default_int_type()),
+                }
+            }
+            Type::Bool => Self::Boolean,
+            Type::String(size) => {
+                let size = size
+                    .evaluate_to_u64()
+                    .expect("Cannot have variable sized strings as a parameter to main");
+                Self::String { length: size }
+            }
+            Type::FmtString(_, _) => unreachable!("format strings cannot be used in the abi"),
+            Type::Error => unreachable!(),
+            Type::Unit => unreachable!(),
+            Type::Constant(_) => unreachable!(),
+            Type::Struct(def, ref args) => {
+                let struct_type = def.borrow();
+                let fields = struct_type.get_fields(args);
+                let fields = vecmap(fields, |(name, typ)| (name, Self::from_type(&typ)));
+                Self::Struct { fields, name: struct_type.name.to_string() }
+            }
+            Type::Tuple(_) => todo!("as_abi_type not yet implemented for tuple types"),
```

Comment:
Nit: as_abi_type is no longer the name of the function where this panic occurs

Reply:
```suggestion
            Type::Tuple(_) => todo!("AbiType::from_type not yet implemented for tuple types"),
```